ubuntu-advantage-client/ec2_cleanup.py

```diff
@@ -4,77 +4,198 @@
 # Copyright 2018 Canonical Ltd.
 # Joshua Powers <josh.powers@canonical.com>
 
-import boto3
-
 import argparse
+import boto3
+import datetime
+import re
 
 
 CI_DEFAULT_TAG = "uaclient-*"
 
+# Our CI pycloudlib use reuses a single shared VPC across multiple
+# CI jobs because VPCs counts are limited to 5 per region.
+# cloud-init has one vpc and uaclient a 2nd shared VPC.
+# If we have any instances running in this VPC, don't try to
+# remove security_groups, subnets, gateways
+SHARED_VPC_TAG = "uaclent-integration"
+
 
 def parse_args():
     parser = argparse.ArgumentParser()
     parser.add_argument(
         "-t", "--tag", dest="tag", action="store", default=CI_DEFAULT_TAG,
-        help=("Tag used to filter cloud resources for deletion. "
+        help=("Tag used to filter cloud resources for deletion by tag. "
               "Default: {}".format(CI_DEFAULT_TAG))
     )
+    parser.add_argument(
+        "-b", "--before-date", dest="before_date", action="store",
+        help=("Resources created before this date will be deleted."
+              " Format: MM/DD/YY")
+    )
     return parser.parse_args()
 
 
-def clean_ec2(tag):
+def get_time_prefix(tag, before_date):
+    """Return time-based tag prefix used to limit deletion to stale resources
+
+    :param tag: String of the general tag filter provided on the commandline.
+    :param before_date: String of the format mm/dd/yyY
+
+    :return: String to be used for further time-based filtering
+    """
+    if not before_date:
+       return ""
+    if not tag:
+        tag = CI_DEFAULT_TAG.replace('*','')
+    if tag[-1] != "-":
+        tag += "-"
+    print('Match only resources created before %s' % before_date)
+    time = datetime.datetime.strptime(before_date, "%m/%d/%y")
+    return time.strftime(tag + "%m%d")
+
+
+def delete_resource_by_tag(resource, tag, time_prefix):
+    """Return whether the resource is older than time_prefix or has tag.
+
+    :param resource: Either a dict or boto3 instance related to a boto3
+        resource. This can be an instance, security_group, subnet etc. SSH keys
+        are processed as dictionaries which contain a KeyName key.
+
+    :param tag: String provided of the generic filter tag provided on the
+        commandline.
+    :param time_prefix: Optional string providing a more specific time filter.
+        When provided, limit deletion to only those resources older than
+        time_prefix.
+
+    If no Name tag present, assume stale and return True
+
+    :return: True if resource should be deleted
+    """
+    if isinstance(resource, dict):
+        if "KeyName" in resource:
+            tag_value = resource["KeyName"]
+    else:
+        for tag in resource.tags:
+            if tag["Key"] != "Name":
+                continue
+            tag_value = tag["Value"]
+
+    if time_prefix:
+        if tag_value >= time_prefix:  # Resource is newer
```

Comment:
Just to be sure here, if we have a tag like `uaclient-*`, time_prefix would be something like `uaclient-*08122020", right ? If that is true, I think this check will fail because of the "*" in the string.

Reply:
Got catch.

I fixed it now.

get_time_prefix("uaclient-*", "08/12/20") will now return 
"uaclient-0812"
instead of 

'uaclient-*-0812'